### ISSUE TITLE
Support for Joroto X4S and Wahoo TICKR

### ISSUE THREAD

**Issue by b202i** [cagnulein/qdomyos-zwift#2903]
**Describe the bug**
Im running QZ version 2.18.9 beta on an Android Pixel 6 Android 14 phone, with a Joroto X4S bluetooth bike and Wahoo TICKR heart band. When configured with all defaults, only the text startup screen is shown, bike ftms is disabled. It should automatically find and select the bike, find and select the heart rate monitor and dynamically display riding telemetry, but it shows the procedural startup screen.

**To Reproduce**
1. Clear data from QZ app, clear app from memory
2. Put on heart strap, batteries are fresh, red-blue lights flashing
3. Start peddling the bike, hit mode to activate bike display, metrics are shown
4. Start QZ, complete the startup survey:
5. Enter age, weight..
6. Select bike from the list
7. Select the TICKR from the list
8. Reach startup text screen

![image](https://github.com/user-attachments/assets/4fd3f13d-3858-46af-aca6-b327ab1e110b)

I’ve also run QZ on a Samsung S10 Ultra with the same result.

**Expected behavior**
 should automatically find and select the bike, find and select the heart rate monitor and dynamically display riding telemetry.

**Screenshots**
Using settings, if I manually select the heart rate monitor and bike, I can get the tile screen showing an active heart rate, and the JORORO bike found, but all other numbers are always zero.
![image](https://github.com/user-attachments/assets/9b20ccbd-0002-4a98-96eb-7163495261e2)

**Desktop (please complete the following information):**
n/a no desktop in this configuration

**Smartphone (please complete the following information):**
 - Device: Google Pixel 6
 - OS: Android
 - Version: 14

**Append a debug log**
[debug-Wed_Dec_11_17_29_47_2024.log](https://github.com/user-attachments/files/18115038/debug-Wed_Dec_11_17_29_47_2024.log)

  
**Additional context**
From the log file debug-Wed_Dec_11_17_29_47_2024.log

- Wed Dec 11 17:29:48 2024 1733956188183 Debug: …/src/main.cpp int main(int, char **) “heart_rate_belt_name” QVariant(QString, “TICKR C0C9”)
- 74 Wed Dec 11 17:29:48 2024 1733956188074 Debug: …/src/main.cpp int main(int, char **) “filter_device” QVariant(QString, “JOROTO-3132”)
- 857 Wed Dec 11 17:29:51 2024 1733956191440 Debug: Device found: “JOROTO-3132” “E2:00:00:00:0C:3C” isLeScanResult: true Manufacturer data size: 0
- 911 Wed Dec 11 17:29:51 2024 1733956191641 Debug: …/src/devices/bluetooth.cpp void bluetooth::debug(const QString &) “deviceUpdated JOROTO-3132 \u0001”
- 986, 988, 1008, 1016, 1059, 1052 (many more, 1 sec apart)
- 1178 Wed Dec 11 17:30:00 2024 1733956200900 Debug: Device found: “JOROTO-3132” “E2:00:00:00:0C:3C” isLeScanResult: true Manufacturer data size: 0
- 680 Wed Dec 11 17:29:50 2024 1733956190527 Warning: qrc:/ChartFooterInnerJS.qml qrc:/ChartFooterInnerJS.qml:17: ReferenceError: rootItem is not defined
- 681 Wed Dec 11 17:29:50 2024 1733956190530 Warning: qrc:/HomeForm.ui.qml qrc:/HomeForm.ui.qml:61: ReferenceError: rootItem is not defined
- 828 Wed Dec 11 17:29:51 2024 1733956191219 Warning: qrc:/Home.qml qrc:/Home.qml:158: ReferenceError: appModel is not defined
-

**Comment by cagnulein** [comment]
Thanks I will check it ASAP

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno gio 12 dic 2024 alle 18:18 MakerMattDesign <
***@***.***> ha scritto:

**Comment by cagnulein** [comment]
good news @b202i the problem is that i didn't add this specific model, but i can add it! Install the free app nrfconnect and show me a screenshot of the bike on the app and share it! Thanks!

**Comment by b202i** [comment]
No problem. Not sure what to capture, so here are a few. I can dig down more if necessary.
![Screenshot_20241213-193205](https://github.com/user-attachments/assets/ebb666cc-9d3d-4097-bc25-8d7e063878c6)
![Screenshot_20241213-193224](https://github.com/user-attachments/assets/753f7957-3572-4e62-b3ff-c521cb92e31c)
![Screenshot_20241213-193234](https://github.com/user-attachments/assets/2dba61b9-ca7d-4c5f-a423-02a471d4c7ff)
![Screenshot_20241213-193243](https://github.com/user-attachments/assets/d30853b3-350d-4dc6-af6e-3baa70013a7c)
![Screenshot_20241213-193305](https://github.com/user-attachments/assets/ed53242a-2b50-40fd-977e-892f68cd31d0)

**Comment by cagnulein** [comment]
Ok perfect, you can set the bike under the accessories, cadence sensor, and also enable the cadence sensor as a bike setting 
Let me know!

**Comment by b202i** [comment]
Setting cadence under accessories worked, it shows on the tile display. All other tile values are zero, including heart-rate. I am unable to find bike under the accessories. Under Bike Options, I continue to leave FTMS Bike as Disabled.

**Comment by cagnulein** [comment]
did you set the heart rate under heart rate belt name setting?
if so could you please show me a fresh debug log?

**Comment by b202i** [comment]
Yes: did you set the heart rate under heart rate belt name setting?

[debug-Sat_Dec_14_17_29_51_2024.log](https://github.com/user-attachments/files/18137985/debug-Sat_Dec_14_17_29_51_2024.log)
[debug-Sat_Dec_14_17_31_04_2024.log](https://github.com/user-attachments/files/18137986/debug-Sat_Dec_14_17_31_04_2024.log)
[debug-Sat_Dec_14_17_31_43_2024.log](https://github.com/user-attachments/files/18137987/debug-Sat_Dec_14_17_31_43_2024.log)
[debug-Sat_Dec_14_17_32_15_2024.log](https://github.com/user-attachments/files/18137988/debug-Sat_Dec_14_17_32_15_2024.log)

**Comment by cagnulein** [comment]
The settings are right but the issue is that the heart rate belt is not advertising itself. This means that it's not awake or it is connected somewhere else. Try to unpair it from the phone

**Comment by b202i** [comment]
With the Wahoo app I can confirm the TICKR C0C9 connects and is the latest firmware. Even though the Wahoo app is picking up a heartbeat, the TICKR C0C9 does not appear as a bluetooth connected device. The TICKR is not visible when pairing new devices. Also of note, the Wahoo app is also able connect the JOROTO bike cadence and speed, although the speed is incorrect, 6 MPH should read about 18 MPH.  It does not appear to receive power/watts from the bike.

From the startup QC page, I can see it cycling through bluetooth devices, including "TICKR C0C9 found". The app can see the device but does not always select it for some reason. I just tried QC 10 times and it did connect to the TICKR a single time, did not connect the other 9. I cleared the app from memory to have a clean start each time. The single time it did connect, I had heart-rate and cadence, but no speed or power.

**Comment by cagnulein** [comment]
Qz required that the device must be in advertising mode: the 9 times that
you don't see it, it was available in the top label? Do you see always with
the nrfconnect app?

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno dom 15 dic 2024 alle 18:14 MakerMattDesign <
***@***.***> ha scritto:

**Comment by b202i** [comment]
I tried nRF Connect several times and it connects to the TICKR and Joroto each time. In Qz I don't always see TICKR in the top label. There is a Toast message that says something like, Bluetooth name is too long. It might not be related to this issue.

![nRF Connect-tickr](https://github.com/user-attachments/assets/aa328f76-dddb-4699-9289-f0b6eb9bda6b)

**Comment by cagnulein** [comment]
In the log i don't see the device advertising. Do you mind share other logs from qz of your tries?

**Comment by b202i** [comment]
Here you go. Thanks again for your help.
[debug-Sun_Dec_15_13_19_30_2024.log](https://github.com/user-attachments/files/18144425/debug-Sun_Dec_15_13_19_30_2024.log)
[debug-Wed_Dec_11_17_29_47_2024.log](https://github.com/user-attachments/files/18144426/debug-Wed_Dec_11_17_29_47_2024.log)

**Comment by cagnulein** [comment]
I can see that once it was connected to it

hrm_lastdevice_address" QVariant(QString, "E9:74:C5:7B:30:A3")

But I can't see advertising anymore

Just to be sure, on nrfconnect, if you clear the cache it always appears? 

I can check if I can use the Mac address to connect to it in case but it's a very non standard behavior from a Bluetooth device

**Comment by b202i** [comment]
I cleared the cash on nrfconnect, and TICKR C0C9 appears after just a couple of seconds. Did it twice, same result each time. The Joroto also shows my heart rate. Can more than one device connect to the TICKR at the same time? As I type this, I am not at the bike but at my desk some 30 feet away. I'll take the batteries out of the bike and report back.

**Comment by cagnulein** [comment]
once the tickr is connected to a device it's stopping for sure to advertise itself. so nope, it shouldn't be connected to different devices.
the fact that you see on nrfconnect everytime it leaves me very stunned. In the QZ log i can see it only once. Unfortunately if it doesn't advertise itself in QZ I can't connect to it. Is your only HR device?

**Comment by cagnulein** [comment]
in nrfconnect don't press connect to it, otherwise you are hiding it from qz

**Comment by b202i** [comment]
Used a second HR monitor, TICKR RDF2, it has older firmware. Now the Joroto computer does not show a HR.  I still get hung up on the instruction page. Devices TICKR 3DF2 and JOROTO 3132 appear in the header as found. While sitting on the bike nrfconnect shows both devices after a few seconds.

Log file and nrfconnect snap attached.
[debug-Mon_Dec_16_11_31_31_2024.log](https://github.com/user-attachments/files/18153832/debug-Mon_Dec_16_11_31_31_2024.log)
![nrf-tickr-3DF2](https://github.com/user-attachments/assets/f923106b-dab0-42fc-bf6d-ff9f2d4683eb)

**Comment by cagnulein** [comment]
You have now to set the joroto as a cadence sensor and also enable the cadence sensor as a bike setting 

Let me know

**Comment by b202i** [comment]
Under Setting-Accessories-Cadence Sensor Options, I was able to select the Joroto but it only worked one time, now the device never shows up. Next, when you change anything in settings the dialog Settings Changed offers to restart the app. One time when I said Yes, if flashed the tile screen while exiting, showing it was full of metrics and had been working. Although when I restarted, it was still no-go. One other thing, I noticed the Joroto computer now displays my HR, using the second HR band. Previously it showed zero for the HR, but now it is active. I think it is finding and attaching to the HR monitor, which prevents QZ from attaching to the HR monitor.

One idea is to buy the Wahoo Cadence sensor and the Wahoo Speed sensor and upgrade the Joroto. I would take the batteries of the Joroto computer and ignore it. Is that a good path forward?

**Comment by cagnulein** [comment]
> Under Setting-Accessories-Cadence Sensor Options, I was able to select the Joroto but it only worked one time, now the device never shows up. Next, when you change anything in settings the dialog Settings Changed offers to restart the app. One time when I said Yes, if flashed the tile screen while exiting, showing it was full of metrics and had been working. Although when I restarted, it was still no-go.

Show me a fresh log now, so I can understand which is your issue. It must work in this way

**Comment by b202i** [comment]
Latest log and settings follow. That is a log of settings!
[debug-Wed_Dec_18_10_05_20_2024.log](https://github.com/user-attachments/files/18184492/debug-Wed_Dec_18_10_05_20_2024.log)
[settings_default_20241218100642.qzs.txt](https://github.com/user-attachments/files/18184501/settings_default_20241218100642.qzs.txt)

**Comment by b202i** [comment]
If you want to modify the settings, I can import them back into the app, or just tell me what to change.

**Comment by cagnulein** [comment]
the problem is the cadence sensor as a bike under the accessories settings
it's disabled. it must be enabled on your case.
let me know

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno mer 18 dic 2024 alle ore 16:17 MakerMattDesign <
***@***.***> ha scritto:

**Comment by b202i** [comment]
Success! ..Um, now I am not sure what other parameters I had changed, there were quite a few restarts along the way. I have the working parameters. I could do a reset, export default parameters, save each as a .txt file, and do a diff to see what had changed. I also need to test with Zwift. 

![success-tiles](https://github.com/user-attachments/assets/949394f6-3396-4376-b0cd-3bc110ae7d3e)

**Comment by cagnulein** [comment]
great! let me know!

**Comment by b202i** [comment]
I'm not sure how it should work. I've had a couple of rides now to learn that the resistance is not being used. The resistance tile remains unchanged when making adjustments. Cadence and heart-rate seem to work perfectly in QZ and Zwift. (ps at end of this post)

Here are most of the diff, lines with '>' are non-defaults

< watt_ignore_builtin=true
> watt_ignore_builtin=false
< filter_device=Disabled
> filter_device=JOROTO-3132
< heart_rate_belt_name=Disabled
> heart_rate_belt_name=TICKR 3DF2
> hrm_lastdevice_address=F8:A9:B5:26:F7:52
> hrm_lastdevice_name=TICKR 3DF2
< bike_resistance_offset=4
> bike_resistance_offset=18
> bluetooth_lastdevice_address=E2:00:00:00:0C:3C
> bluetooth_lastdevice_name=JOROTO-3132
< cadence_sensor_as_bike=false
< cadence_sensor_name=Disabled
> cadence_sensor_as_bike=true
> cadence_sensor_name=JOROTO-3132

[settings-diff.txt](https://github.com/user-attachments/files/18220548/settings-diff.txt)

PS. I reset the app to defaults and am not able to get the heart-rate working. I'll report back after more tinkering.

**Comment by cagnulein** [comment]
Your bike doesn't have resistance control from the app itself, is this what
you mean?
Peloton resistance should change based on your cadence heart rate and
wattage

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno dom 22 dic 2024 alle 04:07 MakerMattDesign <
***@***.***> ha scritto:

**Comment by b202i** [comment]
The bike has manual resistance control. There is a knob you turn that changes the resistance between 0 and 100.

**Comment by cagnulein** [comment]
Ok so the bike don't send the resistance actually so the only thing that I
could do is creating a resistance value from your hr and cadence

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno dom 22 dic 2024 alle 22:56 MakerMattDesign <
***@***.***> ha scritto:

**Comment by b202i** [comment]
Thanks Roberto, I am sure it would help but there would still be a good bit of lag. In Zwift I'm trying to replicate real group riding where I would shift to the big chainring, go anaerobic for a bit and jump across the gap to the next group. There seems to be a very loose connection between cadence and bike speed. Even without resistance delta (or changing gears) if I were to jump on it and change cadence from 60 to 120 rpm I would expect the speed to double. Getting the cadence vs speed relationship is likely possible, but using resistance in place of gear shifting seems to be not possible with this bike.

**Comment by cagnulein** [comment]
Yes because the power it's coming from your heart rate in 90% and cadence for 10% so the lag comes from the fact that the hr needs some tenth of seconds to raise.

On the manufacturer app, is it able to read power or even resistance value?

**Comment by b202i** [comment]
The results have been inconsistent, so I bought a Garmin HRM-Dual chest strap. Will report back with results. The good news is thanks to your help I am at least riding!
